File: BOJ/2019/07/10/10800.py
import sys
total=0
sub_total=0
sub_num=0
S=[0]*220000
D=[]
ar=[]

N=int(sys.stdin.readline())
for i in range(N):
    color,cnt=list(map(int,sys.stdin.readline().split()))
    ar.append([i,color,cnt])
ar.sort(key=lambda t: t[1])
ar.sort(key=lambda t : t[2])

sub_total+=ar[0][2]
S[ar[0][1]]+=ar[0][2]
D.append([ar[0][0],0])

for i in range(1,N):
    if ar[i][2] > ar[i-1][2]:
        total += sub_total + sub_num
        sub_total=0
        sub_num=0
        sub_total += ar[i][2]
        D.append([ar[i][0], total - S[ar[i][1]]])
        S[ar[i][1]] += ar[i][2]
    elif ar[i][2] == ar[i-1][2] and ar[i][1] ==ar[i-1][1]:
        sub_num +=ar[i][2]
        sub_total += ar[i][2]
        D.append([ar[i][0], total])
        S[ar[i][1]] += ar[i][2]
    else:
        total += sub_num
        sub_total += ar[i][2]
        D.append([ar[i][0], total - S[ar[i][1]]])
        S[ar[i][1]] += ar[i][2]

D.sort(key=lambda t : t[0])

for i in D:
    print(i[1])

File: BOJ/10000~10999/10800.cpp
#include <iostream>
#include <algorithm>
#define ll long long
using namespace std;

struct st{
    ll k;
    ll color;
    ll cnt;
};


bool compare(st A, st B){
    if (A.cnt<B.cnt)
    {
        return true;
    }
    else if (A.cnt == B.cnt)
    {
        if(A.color<B.color) return true;
    }
    return false;
}

ll N;
ll total=0,sub_total;
ll S[220000];
ll D[220000];
st ar[220000];

int main()
{
    ios_base :: sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    cin>>N;
    for(int x=0; x<N; x++)
    {
        ar[x].k=x;
        cin>>ar[x].color >>ar[x].cnt;
    }

    sort(ar,ar+N,compare);
    sub_total+=ar[0].cnt;
    D[ar[0].k]=0;

    for(int x=1; x<N; x++)
    {
        if(ar[x].cnt!=ar[x-1].cnt)
        {
            total+=sub_total;
            sub_total=0;
        }
        D[ar[x].k] = total - S[ar[x].color];
        sub_total+=ar[x].cnt;
        S[ar[x].color] += ar[x].cnt;
    }
    
    for(int x=0; x<N; x++){
        cout<<D[x]<<"\n";
    }
    
}